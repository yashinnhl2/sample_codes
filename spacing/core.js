/* eslint-disable dot-notation */
/* eslint-disable import/prefer-default-export */
import { css } from 'styled-components';
import { isArray, isString, size as arraySize } from 'lodash-es';

const getMediaSize = (size, theme) => {
  if (size === 'xs') return theme.mediaXs;
  if (size === 'sm') return theme.mediaSm;
  if (size === 'md') return theme.mediaMd;

  return theme.mediaLg;
};

// This function will convert ['xs', 'md', 'xs' , 'sm'] to `0.25rem 1rem 0.25rem 0.5rem`
// based on array size, it will generate other shorthand syntaxes like vertical and horizontal spacing ['md', 'xs']
const generateShortHandSpacing = (value, theme) => {
  if (arraySize(value) > 4) throw Error('shorthand spacing cannot have more than 4 items');

  return value.reduce((acc, curr) => `${acc + theme.spacing[curr]} `, '');
};

const generateSpacing = ({ value, theme }) => {
  if (isArray(value, theme)) {
    return generateShortHandSpacing(value, theme);
  }

  return `${theme.spacing[value]}`;
};

const getValue = (value, size) => {
  if (isString(value) || isArray(value)) return value;

  return value[size];
};

const generateResponsiveTemplate = (value, cssProperty) => {
  if (!value) return '';

  return css`
    ${(value.xs || isString(value) || isArray(value)) &&
    css`
      ${[cssProperty]}: ${({ theme }) => generateSpacing({ value: getValue(value, 'xs'), theme })};
    `}

    ${value.sm &&
    css`
      @media (min-width: ${({ theme }) => getMediaSize('sm', theme)}) {
        ${[cssProperty]}: ${({ theme }) => generateSpacing({ value: getValue(value, 'sm'), theme })};
      } ;
    `}

    ${value.md &&
    css`
      @media (min-width: ${({ theme }) => getMediaSize('md', theme)}) {
        ${[cssProperty]}: ${({ theme }) => generateSpacing({ value: getValue(value, 'md'), theme })};
      } ;
    `}

    ${value.lg &&
    css`
      @media (min-width: ${({ theme }) => getMediaSize('lg', theme)}) {
        ${[cssProperty]}: ${({ theme }) => generateSpacing({ value: getValue(value, 'lg'), theme })};
      } ;
    `}
  `;
};

// MARGINS
const m = all => {
  return generateResponsiveTemplate(all, 'margin');
};

const mt = top => {
  return generateResponsiveTemplate(top, 'margin-top');
};

const mb = bottom => {
  return generateResponsiveTemplate(bottom, 'margin-bottom');
};

const mr = right => {
  return generateResponsiveTemplate(right, 'margin-right');
};

const ml = left => {
  return generateResponsiveTemplate(left, 'margin-left');
};

const mx = horizontal => {
  return css`
    ${mr(horizontal)};
    ${ml(horizontal)};
  `;
};

const my = vertical => {
  return css`
    ${mt(vertical)};
    ${mb(vertical)};
  `;
};

// PADDINGS
const p = all => {
  return generateResponsiveTemplate(all, 'padding');
};

const pt = top => {
  return generateResponsiveTemplate(top, 'padding-top');
};

const pb = bottom => {
  return generateResponsiveTemplate(bottom, 'padding-bottom');
};

const pr = right => {
  return generateResponsiveTemplate(right, 'padding-right');
};

const pl = left => {
  return generateResponsiveTemplate(left, 'padding-left');
};

const px = horizontal => {
  return css`
    ${pr(horizontal)};
    ${pl(horizontal)};
  `;
};

const py = vertical => {
  return css`
    ${pt(vertical)};
    ${pb(vertical)};
  `;
};

export const Sizes = {
  xs: 'xs',
  sm: 'sm',
  md: 'md',
  lg: 'lg',
  xl: 'xl',
  '2xl': '2xl',
  '3xl': '3xl',
  '4xl': '4xl',
  '5xl': '5xl',
  '6xl': '6xl'
};

/**
 * Pre-defined sizes for margins and padding
 * @typedef {keyof Sizes} CoreSizes
 */

/**
 * An object property used to set properties on specific viewport sizes
 * @typedef {object} ViewportProps
 * @property {CoreSizes} xs - Extra small viewport margin
 * @property {CoreSizes} sm - Small viewport margin
 * @property {CoreSizes} md - Medium viewport margin
 * @property {CoreSizes} lg - Large viewport margin
 */

export { m, mx, my, mt, mr, mb, ml, p, px, py, pt, pr, pb, pl };
