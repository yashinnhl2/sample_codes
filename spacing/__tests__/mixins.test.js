import React from 'react';
import matchesSnapshot from '@upgrade/ui-utils/utils/assertions/matches-snapshot';
import styled from 'styled-components';
import { m, p, mt } from '../core';

describe('Mixins', () => {
  it('Element should get space from all side without queries', () => {
    const Element = styled.div`
      ${m('xl')}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should have get margin from top - no queries', () => {
    const Element = styled.div`
      ${mt('xl')}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should have get padding from all side with all media queries', () => {
    const Element = styled.div`
      ${p({ xs: 'xs', sm: 'sm', md: 'md', lg: 'lg' })}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should have get padding from all side with md media query', () => {
    const Element = styled.div`
      ${p({ md: 'md' })}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should add custom margin to all directions in one line, if array passed', () => {
    const Element = styled.div`
      ${m(['md', 'xs', 'lg', 'md'])}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should add custom margin to horizontal and vertical directions', () => {
    const Element = styled.div`
      ${m(['md', 'xl'])}
    `;
    matchesSnapshot(<Element />);
  });

  it('Should add custom margin to horizontal and vertical directions in different media queries', () => {
    const Element = styled.div`
      ${m({ xs: ['md', 'xs'], sm: ['xl', 'md'] })}
    `;
    matchesSnapshot(<Element />);
  });
});
