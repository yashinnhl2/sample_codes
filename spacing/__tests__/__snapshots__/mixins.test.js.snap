// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mixins Element should get space from all side without queries 1`] = `
.c0 {
  margin: 2rem;
}

<div
  className="c0"
/>
`;

exports[`Mixins Should add custom margin to all directions in one line, if array passed 1`] = `
.c0 {
  margin: 1rem 0.25rem 1.5rem 1rem;
}

<div
  className="c0"
/>
`;

exports[`Mixins Should add custom margin to horizontal and vertical directions 1`] = `
.c0 {
  margin: 1rem 2rem;
}

<div
  className="c0"
/>
`;

exports[`Mixins Should add custom margin to horizontal and vertical directions in different media queries 1`] = `
.c0 {
  margin: 1rem 0.25rem;
}

@media (min-width:48em) {
  .c0 {
    margin: 2rem 1rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Mixins Should have get margin from top - no queries 1`] = `
.c0 {
  margin-top: 2rem;
}

<div
  className="c0"
/>
`;

exports[`Mixins Should have get padding from all side with all media queries 1`] = `
.c0 {
  padding: 0.25rem;
}

@media (min-width:48em) {
  .c0 {
    padding: 0.5rem;
  }
}

@media (min-width:64em) {
  .c0 {
    padding: 1rem;
  }
}

@media (min-width:75em) {
  .c0 {
    padding: 1.5rem;
  }
}

<div
  className="c0"
/>
`;

exports[`Mixins Should have get padding from all side with md media query 1`] = `
@media (min-width:64em) {
  .c0 {
    padding: 1rem;
  }
}

<div
  className="c0"
/>
`;
